### ISSUE TITLE
(Lack of) Symmetry in Interaction Terms

### ISSUE THREAD

**Issue by palday** [JuliaStats/StatsModels.jl#198]
Right now, interaction terms are constructed with their constituent bits in the order specified in the formula:

```julia

julia> ab = @formula(0 ~ a * b)
FormulaTerm
Response:
  0
Predictors:
  a(unknown)
  b(unknown)
  a(unknown) & b(unknown)

julia> ba = @formula(0 ~ b * a)
FormulaTerm
Response:
  0
Predictors:
  b(unknown)
  a(unknown)
  b(unknown) & a(unknown)
``` 
The equality of `InteractionTerms` is impacted by this behavior:

```julia
julia> ba.rhs[2] == ab.rhs[1] # the terms for b are equal
true

julia> ba.rhs[3] == ab.rhs[3] # the terms for 
false
```

I can't find anything in the docs about whether this is behavior is defined to be this way or whether it's just an implementation detail. For my own purposes, it would be swell if there were a canonical ordering to interaction terms (especially since there is an lexical ordering applied in other contexts, e.g. the order of levels of a categorical term), but I can see counter arguments for keeping the order as specified in the formula. No matter which way the decision falls, I think it should be documented (or that the ordering is an implementation detail and subject to change).

The other side of this coin is that `a&b != b&a`. For standard use of interaction terms in statistics, this symmetry holds because they're formed by the element-wise multiplication of the two columns, which is a commutative operation. It's probably very difficult to make this equivalence hold without having a canonical ordering somewhere, whether if it's in a stored private field or recomputed on the fly.  Again, I can see some potential counter arguments to this, especially for more customized schema, but we should probably make an explicit documentation note because this goes against the usual interpretation and underlying math of interactions.

cc: @ararslan

**Comment by kleinschmidt** [comment]
Yeah it's tricky.  Because the order that the sub-terms occur will determine the order of the columns that are generated (and possibly which if any columns get dropped by GLM/MixedModels if the matrix is rank deficient...), so I'm hesitant to make them actually equal.  Maybe we could add some kind of "clean-up" pass which sorts them lexically as well (or even a method for `Base.sort`).
